### DISCUSSION TITLE
Is there a `HashMap` type?

### DISCUSSION THREAD

**Discussion by TTWNO** [wasmerio/wasmer#3848] [General]
I would like to know if there is a way to generate a `HashMap<K, V>` in Rust?

Pretty sure it could be emulated with `list<tuple<K, V>>`, but not sure if that's recognized as a special case or not.

Thanks,

**Comment by Michael-F-Bryan** [comment]
For now, I would use a `list<tuple<K, V>>` or `tuple<list<K>, list<V>>`[^1] to transfer key-value pairs between the host and guest, then reassemble the HashMap manually.

I've created a ticket on the WAI repo to track this - https://github.com/wasmerio/wai/issues/40.

[^1]: they're roughly equivalent, but the former is more convenient to work with, while the latter can be slightly more efficient to transfer because `tuple<K, V>` might add padding.
